boost format operator % won't accept dereference of volatile

Originally reported by [Paul Rose in Boost Trac](https://svn.boost.org/trac10/ticket/10852).

```
struct A{
    A() : a(9) {} int a;
}; A a; A volatile *pva = &a;

std::cout << boost::format("%d") % pva->a << std::endl;

This worked in 1_54 but fails in 1_56 with error:
    format/feed_args.hpp:135:47: error: invalid conversion from 'volatile void*' to 'const void*'

Can work around this with cast:
    std::cout << boost::format("%d") % (int)pva->a << std::endl;
but should not have to. The appearance to the user is pass by value.

Running linux with g++ 4.6.2 and 4.8.2
```

Originally reported in Boost 1.56.0.
Confirmed in Boost 1.59.0 with clang-3.8 on linux.
Confirmed in Boost 1.61.0 with VS2008.
Confirmed in Boost 1.63.0 with gcc-5.2.

--- a/include/boost/format/format_class.hpp
+++ b/include/boost/format/format_class.hpp
@@ -68,6 +68,18 @@ namespace boost {
             { return io::detail::feed<CharT, Tr, Alloc, T&>(*this,x); }
 #endif
 
+        template<class T>
+        basic_format& operator%(volatile const T& x)
+            { /* make a non-volatile copy */ const T v(x);
+              /* pass the copy along      */ return io::detail::feed<CharT, Tr, Alloc, const T&>(*this, v); }
+
+#ifndef BOOST_NO_OVERLOAD_FOR_NON_CONST
+        template<class T>
+        basic_format& operator%(volatile T& x)
+            { /* make a non-volatile copy */ T v(x);
+              /* pass the copy along      */ return io::detail::feed<CharT, Tr, Alloc, T&>(*this, v); }
+#endif
+
 #if defined(__GNUC__)
         // GCC can't handle anonymous enums without some help
         // ** arguments passing ** //

--- a/test/format_test2.cpp
+++ b/test/format_test2.cpp
@@ -186,5 +186,11 @@ int test_main(int, char* [])
     BOOST_CHECK_THROW(boost::format("%I63d"), boost::io::bad_format_string);
     BOOST_CHECK_THROW(boost::format("%I128d"), boost::io::bad_format_string);
 
+    // issue-36 volatile (and const) keyword
+    volatile int vint = 1234567;
+    BOOST_CHECK_EQUAL((boost::format("%1%") % vint).str(), "1234567");
+    volatile const int vcint = 7654321;
+    BOOST_CHECK_EQUAL((boost::format("%1%") % vcint).str(), "7654321");
+
     return 0;
 }

